// bindable-decor-bindings/lib/index.js
// Generated by CoffeeScript 1.7.1
(function() {
  var BindingsDecorator, disposable,
    __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };

  disposable = require("disposable");

  BindingsDecorator = (function() {

    /*
     */
    function BindingsDecorator(target, options) {
      this.target = target;
      this.dispose = __bind(this.dispose, this);
      this._onVisibilityChange = __bind(this._onVisibilityChange, this);
      this.bind = __bind(this.bind, this);
      this.bindings = typeof options === "object" ? options : void 0;
      this._bindings = [];
      this.target.once("dispose", this.dispose);
      this.target.on("change:visible", this._onVisibilityChange);
    }


    /*
     */

    BindingsDecorator.prototype.bind = function() {
      this.dispose();
      if (this.bindings) {
        return this._setupExplicitBindings();
      }
    };


    /*
     */

    BindingsDecorator.prototype._onVisibilityChange = function(value) {
      if (value) {
        return this.resume();
      } else {
        return this.pause();
      }
    };


    /*
     */

    BindingsDecorator.prototype.dispose = function() {
      var binding, _i, _len, _ref;
      _ref = this._bindings;
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        binding = _ref[_i];
        binding.dispose();
      }
      return this._bindings = [];
    };


    /*
     */

    BindingsDecorator.prototype.pause = function() {
      var binding, _i, _len, _ref, _results;
      _ref = this._bindings;
      _results = [];
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        binding = _ref[_i];
        _results.push(binding.pause());
      }
      return _results;
    };


    /*
     */

    BindingsDecorator.prototype.resume = function() {
      var binding, _i, _len, _ref, _results;
      _ref = this._bindings;
      _results = [];
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        binding = _ref[_i];
        _results.push(binding.resume().now());
      }
      return _results;
    };


    /*
     explicit bindings are properties from & to properties of the view controller
     */

    BindingsDecorator.prototype._setupExplicitBindings = function() {
      var bindings, key, _results;
      bindings = this.bindings;
      this._bindings = [];
      _results = [];
      for (key in bindings) {
        _results.push(this._setupBinding(key, bindings[key]));
      }
      return _results;
    };


    /*
     */

    BindingsDecorator.prototype._setupBinding = function(property, to) {
      var oldTo, options;
      options = {};
      if (typeof to === "function") {
        oldTo = to;
        to = (function(_this) {
          return function() {
            return oldTo.apply(_this.target, arguments);
          };
        })(this);
      }
      if (to.to) {
        options = to;
      } else {
        options = {
          to: to
        };
      }
      return this._bindings.push(this.target.bind(property, options).now());
    };

    return BindingsDecorator;

  })();

  module.exports = function(event) {
    return {
      priority: "load",
      multi: true,
      getOptions: function(target) {
        return target.bindings;
      },
      decorate: function(target, options) {
        var decor;
        decor = new BindingsDecorator(target, options);
        if (event) {
          return target.once(event, decor.bind);
        } else {
          return decor.bind();
        }
      }
    };
  };

}).call(this);
